core/shared/src/main/scala/zio/internal/LockedRef.scala

```diff
@@ -0,0 +1,39 @@
+package zio.internal
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+class LockedRef[A](var ref: A) {
+  var locked = new AtomicBoolean(false)
+
+  def get: A = ref
+
+  def set(a: A): A = {
+    ref = a
+    ref
+  }
+
+  def synchronized(f: => Unit): Unit = {
+    lock()
+    try f
+    finally unlock()
+  }
+
+  def lock(): Unit = {
+    var loop = true
+    while (loop) {
```

Comment:
This spin loop is not efficient. You should just use an ordinary Java `Semaphore` or `Lock`. An alternative would be to implement a lock atop `synchronized`, which would be more efficient than anything else. You have to know how to use `waitAll` / `notifyAll` and deal with spurious wakeups, etc. So the simple solution is to use `Semaphore` here the more complex one is to build something low-level

Reply:
Thank you for your advice on locking!
Unfortunately, `java.util.concurrent.Semaphore` is not supported by scala.js...
Maybe I should implement lock with `ReentrantLock` here while learning `waitAll` / `notifyAll` and other concurrent matter to be cared for. (java's built-in `synchronized` is not supported as you taught us in https://github.com/zio/zio/issues/803)

But on reading your overall review on this PR, this class is not needed in the first place...?

Reply:
Javascript will not need a semaphore. I think we can get by with purely `synchronized` (and `Sync` class above), without any waiting / awaiting.

Can you give it a try and let me know? Thank you!

Reply:
Yes, you're totally correct.
The mechanism of Javascript runtime slipped my mind.